DigitalLearningSolutions.Web/ServiceFilter/ValidateAllowedDlsSubApplication.cs

```diff
@@ -0,0 +1,134 @@
+            {
+                SetNotFoundResult(context);
+                return;
+            }
+
+            var application = dlsSubApplicationParameterName == null
+                ? null
+                : GetDlsSubApplicationFromContext(context, dlsSubApplicationParameterName);
+
+            if (DlsSubApplication.TrackingSystem.Equals(application) &&
+                !await featureManager.IsEnabledAsync(FeatureFlags.RefactoredTrackingSystem) ||
+                DlsSubApplication.SuperAdmin.Equals(application) &&
+                !await featureManager.IsEnabledAsync(FeatureFlags.RefactoredSuperAdminInterface))
+            {
+                SetNotFoundResult(context);
+                return;
+            }
+
+            if (validApplications.Any() && !validApplications.Contains(application))
+            {
+                SetNotFoundResult(context);
+                return;
+            }
+
+            if (user.IsDelegateOnlyAccount() && DlsSubApplication.Main.Equals(application))
+            {
+                RedirectToLearningPortalVersion(context, dlsSubApplicationParameterName!);
+                return;
+            }
+
+            if (!user.HasLearningPortalPermissions() && DlsSubApplication.LearningPortal.Equals(application) ||
+                !user.HasFrameworksAdminPermissions() && DlsSubApplication.Frameworks.Equals(application) ||
+                !user.HasSupervisorAdminPermissions() && DlsSubApplication.Supervisor.Equals(application) ||
+                !user.HasCentreAdminPermissions() && DlsSubApplication.TrackingSystem.Equals(application) ||
+                !user.HasSuperAdminPermissions() && DlsSubApplication.SuperAdmin.Equals(application))
+            {
+                RedirectToAccessDenied(context);
+            }
+        }
+
+        private bool HasModelBindingError(ActionExecutingContext context, string applicationArgumentName)
+        {
+            return context.ModelState.GetValidationState(applicationArgumentName) == ModelValidationState.Invalid;
+        }
+
+        private void RedirectToLogin(ActionExecutingContext context)
+        {
+            context.Result = new RedirectToActionResult(
+                nameof(LoginController.Index),
+                ControllerHelper.GetControllerAspName(typeof(LoginController)),
+                new { }
+            );
+        }
+
+        private void RedirectToLearningPortalVersion(ActionExecutingContext context, string applicationArgumentName)
+        {
+            var descriptor = ((ControllerBase)context.Controller).ControllerContext.ActionDescriptor;
+            var routeValues = new Dictionary<string, object>
+            {
+                [applicationArgumentName] = DlsSubApplication.LearningPortal,
+            };
+            context.Result = new RedirectToActionResult(descriptor.ActionName, descriptor.ControllerName, routeValues);
+        }
+
+        private void RedirectToAccessDenied(ActionExecutingContext context)

```

Comment:
This method does the same kind of thing as the one below. Can we have similar names for them? I prefer `Set...Result` because `RedirectTo...` looks synchronous.